// @cardinal/creator-standard/dist/cjs/generated/instructions/ApproveAndSetInUseBy.js
"use strict";
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createApproveAndSetInUseByInstruction = exports.approveAndSetInUseByInstructionDiscriminator = exports.ApproveAndSetInUseByStruct = void 0;
const tslib_1 = require("tslib");
const splToken = tslib_1.__importStar(require("@solana/spl-token"));
const beet = tslib_1.__importStar(require("@metaplex-foundation/beet"));
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
const ApproveAndSetInUseByIx_1 = require("../types/ApproveAndSetInUseByIx");
/**
 * @category Instructions
 * @category ApproveAndSetInUseBy
 * @category generated
 */
exports.ApproveAndSetInUseByStruct = new beet.BeetArgsStruct([
    ['instructionDiscriminator', beet.u8],
    ['approveAndSetInUseByIx', ApproveAndSetInUseByIx_1.approveAndSetInUseByIxBeet],
], 'ApproveAndSetInUseByInstructionArgs');
exports.approveAndSetInUseByInstructionDiscriminator = 8;
/**
 * Creates a _ApproveAndSetInUseBy_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category ApproveAndSetInUseBy
 * @category generated
 */
function createApproveAndSetInUseByInstruction(accounts, args, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    var _a;
    const [data] = exports.ApproveAndSetInUseByStruct.serialize({
        instructionDiscriminator: exports.approveAndSetInUseByInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.mintManager,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.ruleset,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.mint,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.inUseByAddress,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.holderTokenAccount,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.holder,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.delegate,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: (_a = accounts.tokenProgram) !== null && _a !== void 0 ? _a : splToken.TOKEN_PROGRAM_ID,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
exports.createApproveAndSetInUseByInstruction = createApproveAndSetInUseByInstruction;
