// @cardinal/creator-standard/dist/cjs/generated/instructions/Burn.js
"use strict";
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createBurnInstruction = exports.burnInstructionDiscriminator = exports.BurnStruct = void 0;
const tslib_1 = require("tslib");
const splToken = tslib_1.__importStar(require("@solana/spl-token"));
const beet = tslib_1.__importStar(require("@metaplex-foundation/beet"));
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
/**
 * @category Instructions
 * @category Burn
 * @category generated
 */
exports.BurnStruct = new beet.BeetArgsStruct([['instructionDiscriminator', beet.u8]], 'BurnInstructionArgs');
exports.burnInstructionDiscriminator = 9;
/**
 * Creates a _Burn_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category Burn
 * @category generated
 */
function createBurnInstruction(accounts, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    var _a, _b;
    const [data] = exports.BurnStruct.serialize({
        instructionDiscriminator: exports.burnInstructionDiscriminator,
    });
    const keys = [
        {
            pubkey: accounts.mintManager,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.mint,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.holderTokenAccount,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.holder,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: (_a = accounts.tokenProgram) !== null && _a !== void 0 ? _a : splToken.TOKEN_PROGRAM_ID,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: (_b = accounts.systemProgram) !== null && _b !== void 0 ? _b : web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
exports.createBurnInstruction = createBurnInstruction;
