// @cardinal/creator-standard/dist/cjs/generated/instructions/CloseRuleset.js
"use strict";
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCloseRulesetInstruction = exports.closeRulesetInstructionDiscriminator = exports.CloseRulesetStruct = void 0;
const tslib_1 = require("tslib");
const beet = tslib_1.__importStar(require("@metaplex-foundation/beet"));
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
/**
 * @category Instructions
 * @category CloseRuleset
 * @category generated
 */
exports.CloseRulesetStruct = new beet.BeetArgsStruct([['instructionDiscriminator', beet.u8]], 'CloseRulesetInstructionArgs');
exports.closeRulesetInstructionDiscriminator = 2;
/**
 * Creates a _CloseRuleset_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category CloseRuleset
 * @category generated
 */
function createCloseRulesetInstruction(accounts, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    const [data] = exports.CloseRulesetStruct.serialize({
        instructionDiscriminator: exports.closeRulesetInstructionDiscriminator,
    });
    const keys = [
        {
            pubkey: accounts.ruleset,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.authority,
            isWritable: true,
            isSigner: true,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
exports.createCloseRulesetInstruction = createCloseRulesetInstruction;
