// @cardinal/creator-standard/dist/cjs/generated/instructions/InitRuleset.js
"use strict";
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInitRulesetInstruction = exports.initRulesetInstructionDiscriminator = exports.InitRulesetStruct = void 0;
const tslib_1 = require("tslib");
const beet = tslib_1.__importStar(require("@metaplex-foundation/beet"));
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
const InitRulesetIx_1 = require("../types/InitRulesetIx");
/**
 * @category Instructions
 * @category InitRuleset
 * @category generated
 */
exports.InitRulesetStruct = new beet.FixableBeetArgsStruct([
    ['instructionDiscriminator', beet.u8],
    ['initRulesetIx', InitRulesetIx_1.initRulesetIxBeet],
], 'InitRulesetInstructionArgs');
exports.initRulesetInstructionDiscriminator = 0;
/**
 * Creates a _InitRuleset_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category InitRuleset
 * @category generated
 */
function createInitRulesetInstruction(accounts, args, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    var _a;
    const [data] = exports.InitRulesetStruct.serialize({
        instructionDiscriminator: exports.initRulesetInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.ruleset,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.authority,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.payer,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: (_a = accounts.systemProgram) !== null && _a !== void 0 ? _a : web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
exports.createInitRulesetInstruction = createInitRulesetInstruction;
