// @cardinal/creator-standard/dist/cjs/generated/instructions/RemoveInUseBy.js
"use strict";
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRemoveInUseByInstruction = exports.removeInUseByInstructionDiscriminator = exports.RemoveInUseByStruct = void 0;
const tslib_1 = require("tslib");
const beet = tslib_1.__importStar(require("@metaplex-foundation/beet"));
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
/**
 * @category Instructions
 * @category RemoveInUseBy
 * @category generated
 */
exports.RemoveInUseByStruct = new beet.BeetArgsStruct([['instructionDiscriminator', beet.u8]], 'RemoveInUseByInstructionArgs');
exports.removeInUseByInstructionDiscriminator = 6;
/**
 * Creates a _RemoveInUseBy_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category RemoveInUseBy
 * @category generated
 */
function createRemoveInUseByInstruction(accounts, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    const [data] = exports.RemoveInUseByStruct.serialize({
        instructionDiscriminator: exports.removeInUseByInstructionDiscriminator,
    });
    const keys = [
        {
            pubkey: accounts.mintManager,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.user,
            isWritable: false,
            isSigner: true,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
exports.createRemoveInUseByInstruction = createRemoveInUseByInstruction;
