// @cardinal/creator-standard/dist/cjs/generated/instructions/SetInUseBy.js
"use strict";
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSetInUseByInstruction = exports.setInUseByInstructionDiscriminator = exports.SetInUseByStruct = void 0;
const tslib_1 = require("tslib");
const beet = tslib_1.__importStar(require("@metaplex-foundation/beet"));
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
/**
 * @category Instructions
 * @category SetInUseBy
 * @category generated
 */
exports.SetInUseByStruct = new beet.BeetArgsStruct([['instructionDiscriminator', beet.u8]], 'SetInUseByInstructionArgs');
exports.setInUseByInstructionDiscriminator = 5;
/**
 * Creates a _SetInUseBy_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetInUseBy
 * @category generated
 */
function createSetInUseByInstruction(accounts, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    const [data] = exports.SetInUseByStruct.serialize({
        instructionDiscriminator: exports.setInUseByInstructionDiscriminator,
    });
    const keys = [
        {
            pubkey: accounts.mintManager,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.ruleset,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.inUseByAddress,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.holder,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.holderTokenAccount,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
exports.createSetInUseByInstruction = createSetInUseByInstruction;
