// @cardinal/creator-standard/dist/cjs/generated/instructions/UpdateRuleset.js
"use strict";
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUpdateRulesetInstruction = exports.updateRulesetInstructionDiscriminator = exports.UpdateRulesetStruct = void 0;
const tslib_1 = require("tslib");
const beet = tslib_1.__importStar(require("@metaplex-foundation/beet"));
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
const UpdateRulesetIx_1 = require("../types/UpdateRulesetIx");
/**
 * @category Instructions
 * @category UpdateRuleset
 * @category generated
 */
exports.UpdateRulesetStruct = new beet.FixableBeetArgsStruct([
    ['instructionDiscriminator', beet.u8],
    ['updateRulesetIx', UpdateRulesetIx_1.updateRulesetIxBeet],
], 'UpdateRulesetInstructionArgs');
exports.updateRulesetInstructionDiscriminator = 1;
/**
 * Creates a _UpdateRuleset_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category UpdateRuleset
 * @category generated
 */
function createUpdateRulesetInstruction(accounts, args, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    var _a;
    const [data] = exports.UpdateRulesetStruct.serialize({
        instructionDiscriminator: exports.updateRulesetInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.ruleset,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.authority,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.payer,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: (_a = accounts.systemProgram) !== null && _a !== void 0 ? _a : web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
exports.createUpdateRulesetInstruction = createUpdateRulesetInstruction;
