// @cardinal/creator-standard/dist/cjs/generated/instructions/UpdateMintManager.js
"use strict";
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUpdateMintManagerInstruction = exports.updateMintManagerInstructionDiscriminator = exports.UpdateMintManagerStruct = void 0;
const tslib_1 = require("tslib");
const beet = tslib_1.__importStar(require("@metaplex-foundation/beet"));
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
const UpdateMintManagerIx_1 = require("../types/UpdateMintManagerIx");
/**
 * @category Instructions
 * @category UpdateMintManager
 * @category generated
 */
exports.UpdateMintManagerStruct = new beet.BeetArgsStruct([
    ['instructionDiscriminator', beet.u8],
    ['updateMintManagerIx', UpdateMintManagerIx_1.updateMintManagerIxBeet],
], 'UpdateMintManagerInstructionArgs');
exports.updateMintManagerInstructionDiscriminator = 4;
/**
 * Creates a _UpdateMintManager_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category UpdateMintManager
 * @category generated
 */
function createUpdateMintManagerInstruction(accounts, args, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    var _a;
    const [data] = exports.UpdateMintManagerStruct.serialize({
        instructionDiscriminator: exports.updateMintManagerInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.mintManager,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.ruleset,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.authority,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.payer,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: (_a = accounts.systemProgram) !== null && _a !== void 0 ? _a : web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
exports.createUpdateMintManagerInstruction = createUpdateMintManagerInstruction;
