// @cardinal/namespaces-components/dist/cjs/utils/format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tryGetProfile = exports.tryGetImageUrl = exports.formatShortAddress = exports.tryPublicKey = exports.shortPubKey = exports.formatTwitterLink = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("@emotion/react/jsx-runtime");
const namespaces_1 = require("@cardinal/namespaces");
const web3_js_1 = require("@solana/web3.js");
const api_1 = require("./api");
const formatTwitterLink = (handle) => {
    if (!handle)
        return (0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, {});
    return ((0, jsx_runtime_1.jsx)("a", Object.assign({ href: `https://twitter.com/${handle}`, style: { color: "#177ddc" }, target: "_blank", rel: "noreferrer" }, { children: handle })));
};
exports.formatTwitterLink = formatTwitterLink;
function shortPubKey(pubkey) {
    if (!pubkey)
        return "";
    return `${pubkey === null || pubkey === void 0 ? void 0 : pubkey.toString().substring(0, 4)}..${pubkey === null || pubkey === void 0 ? void 0 : pubkey.toString().substring((pubkey === null || pubkey === void 0 ? void 0 : pubkey.toString().length) - 4)}`;
}
exports.shortPubKey = shortPubKey;
const tryPublicKey = (publicKeyString) => {
    if (publicKeyString instanceof web3_js_1.PublicKey)
        return publicKeyString;
    if (!publicKeyString)
        return null;
    try {
        return new web3_js_1.PublicKey(publicKeyString);
    }
    catch (e) {
        return null;
    }
};
exports.tryPublicKey = tryPublicKey;
const formatShortAddress = (address) => {
    if (!address)
        return (0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, {});
    return ((0, jsx_runtime_1.jsx)("a", Object.assign({ href: `https://explorer.solana.com/address/${address.toString()}`, target: "_blank", rel: "noopener noreferrer" }, { children: (0, namespaces_1.shortenAddress)(address.toString()) })));
};
exports.formatShortAddress = formatShortAddress;
function tryGetImageUrl(handle, dev) {
    var _a;
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        try {
            const response = yield fetch(`${(0, api_1.apiBase)(dev)}/twitter/proxy?url=https://api.twitter.com/2/users/by&usernames=${handle}&user.fields=profile_image_url`);
            const json = (yield response.json());
            return (_a = json === null || json === void 0 ? void 0 : json.data[0]) === null || _a === void 0 ? void 0 : _a.profile_image_url.replace("_normal", "");
        }
        catch (e) {
            console.log(e);
            return undefined;
        }
    });
}
exports.tryGetImageUrl = tryGetImageUrl;
function tryGetProfile(handle, dev) {
    var _a, _b, _c, _d;
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        try {
            const response = yield fetch(`${(0, api_1.apiBase)(dev)}/twitter/proxy?url=https://api.twitter.com/2/users/by&usernames=${handle}&user.fields=profile_image_url`);
            const json = (yield response.json());
            return {
                profile_image_url: (_a = json === null || json === void 0 ? void 0 : json.data[0]) === null || _a === void 0 ? void 0 : _a.profile_image_url.replace("_normal", ""),
                username: (_b = json === null || json === void 0 ? void 0 : json.data[0]) === null || _b === void 0 ? void 0 : _b.username,
                id: (_c = json === null || json === void 0 ? void 0 : json.data[0]) === null || _c === void 0 ? void 0 : _c.id,
                name: (_d = json === null || json === void 0 ? void 0 : json.data[0]) === null || _d === void 0 ? void 0 : _d.name,
            };
        }
        catch (e) {
            return undefined;
        }
    });
}
exports.tryGetProfile = tryGetProfile;
