// @cardinal/namespaces-components/dist/esm/components/ProfileSmall.js
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
import { HiUserCircle } from 'react-icons/hi';
import { shortPubKey } from '../utils/format';
import { AddressImage } from './AddressImage';
import { DisplayAddress } from './DisplayAddress';
export const ProfileSmall = ({ connection, address, dark, onClick, placeholder, className, style, }) => {
    return (_jsxs("div", Object.assign({ className: `${className} flex cursor-pointer gap-2 text-sm`, style: style, onClick: onClick }, { children: [_jsx(AddressImage, { connection: connection, address: address || undefined, height: "40px", width: "40px", dark: dark, placeholder: placeholder || (_jsx("div", Object.assign({ style: {
                        color: 'rgb(209, 213, 219)',
                        cursor: 'pointer',
                        overflow: 'hidden',
                        height: '40px',
                        width: '40px',
                    } }, { children: _jsx(HiUserCircle, { style: {
                            position: 'relative',
                            height: '50px',
                            width: '50px',
                            left: '-5px',
                            top: '-5px',
                        }, className: "relative left-[-5px] top-[-5px] h-[50px] w-[50px]" }) }))) }), _jsxs("div", { children: [_jsx("div", Object.assign({ className: `text-${dark ? 'white' : 'black'}` }, { children: _jsx(DisplayAddress, { style: { pointerEvents: 'none' }, connection: connection, address: address || undefined, height: "20px", width: "100px", dark: dark }) })), _jsx("div", Object.assign({ className: "text-sm text-gray-500" }, { children: shortPubKey(address) }))] })] })));
};
