// @metaplex-foundation/mpl-token-auth-rules/dist/src/generated/instructions/Validate.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createValidateInstruction = exports.validateInstructionDiscriminator = exports.ValidateStruct = void 0;
const beet = __importStar(require("@metaplex-foundation/beet"));
const web3 = __importStar(require("@solana/web3.js"));
const ValidateArgs_1 = require("../types/ValidateArgs");
exports.ValidateStruct = new beet.FixableBeetArgsStruct([
    ['instructionDiscriminator', beet.u8],
    ['validateArgs', ValidateArgs_1.validateArgsBeet],
], 'ValidateInstructionArgs');
exports.validateInstructionDiscriminator = 1;
function createValidateInstruction(accounts, args, programId = new web3.PublicKey('auth9SigNpDKz4sJJ1DfCTuZrZNSAgh9sFD3rboVmgg')) {
    var _a, _b, _c, _d;
    const [data] = exports.ValidateStruct.serialize({
        instructionDiscriminator: exports.validateInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.ruleSetPda,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.mint,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: (_a = accounts.systemProgram) !== null && _a !== void 0 ? _a : web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: (_b = accounts.payer) !== null && _b !== void 0 ? _b : programId,
            isWritable: accounts.payer != null,
            isSigner: accounts.payer != null,
        },
        {
            pubkey: (_c = accounts.ruleAuthority) !== null && _c !== void 0 ? _c : programId,
            isWritable: false,
            isSigner: accounts.ruleAuthority != null,
        },
        {
            pubkey: (_d = accounts.ruleSetStatePda) !== null && _d !== void 0 ? _d : programId,
            isWritable: accounts.ruleSetStatePda != null,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
exports.createValidateInstruction = createValidateInstruction;
