// @saberhq/use-solana/dist/cjs/providers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_WALLET_PROVIDERS = exports.DefaultWalletType = void 0;
const wallet_adapter_clover_1 = require("@solana/wallet-adapter-clover");
const wallet_adapter_coin98_1 = require("@solana/wallet-adapter-coin98");
const wallet_adapter_glow_1 = require("@solana/wallet-adapter-glow");
const wallet_adapter_huobi_1 = require("@solana/wallet-adapter-huobi");
const wallet_adapter_mathwallet_1 = require("@solana/wallet-adapter-mathwallet");
const wallet_adapter_phantom_1 = require("@solana/wallet-adapter-phantom");
const wallet_adapter_slope_1 = require("@solana/wallet-adapter-slope");
const wallet_adapter_solflare_1 = require("@solana/wallet-adapter-solflare");
const wallet_adapter_sollet_1 = require("@solana/wallet-adapter-sollet");
const wallet_adapter_solong_1 = require("@solana/wallet-adapter-solong");
const adapters_1 = require("./adapters");
const readonly_1 = require("./adapters/readonly");
const secret_key_1 = require("./adapters/secret-key");
const solflare_1 = require("./adapters/solflare");
const icons_1 = require("./icons");
var DefaultWalletType;
(function (DefaultWalletType) {
    DefaultWalletType["Clover"] = "Clover";
    DefaultWalletType["Coin98"] = "Coin98";
    DefaultWalletType["Ledger"] = "Ledger";
    DefaultWalletType["Glow"] = "Glow";
    DefaultWalletType["MathWallet"] = "MathWallet";
    DefaultWalletType["Phantom"] = "Phantom";
    DefaultWalletType["ReadOnly"] = "ReadOnly";
    DefaultWalletType["Slope"] = "Slope";
    DefaultWalletType["Sollet"] = "Sollet";
    DefaultWalletType["SolletExtension"] = "SolletExtension";
    DefaultWalletType["Solflare"] = "Solflare";
    DefaultWalletType["SolflareExtension"] = "SolflareExtension";
    DefaultWalletType["Solong"] = "Solong";
    DefaultWalletType["SecretKey"] = "SecretKey";
    DefaultWalletType["Huobi"] = "Huobi";
})(DefaultWalletType = exports.DefaultWalletType || (exports.DefaultWalletType = {}));
exports.DEFAULT_WALLET_PROVIDERS = {
    [DefaultWalletType.Sollet]: {
        name: "Sollet",
        url: "https://www.sollet.io",
        icon: icons_1.SOLLET,
        makeAdapter: (provider, network) => new adapters_1.SolanaWalletAdapter(new wallet_adapter_sollet_1.SolletWalletAdapter({
            provider,
            network: network,
        })),
        isMobile: true,
    },
    [DefaultWalletType.SolletExtension]: {
        name: "Sollet (Extension)",
        url: "https://chrome.google.com/webstore/detail/sollet/fhmfendgdocmcbmfikdcogofphimnkno",
        icon: icons_1.SOLLET,
        makeAdapter: (_provider, network) => new adapters_1.SolanaWalletAdapter(new wallet_adapter_sollet_1.SolletExtensionWalletAdapter({
            network: network,
        })),
        isInstalled: () => window.sollet !== undefined,
    },
    [DefaultWalletType.Ledger]: {
        name: "Ledger",
        url: "https://www.ledger.com",
        icon: icons_1.LEDGER,
        makeAdapter: () => new adapters_1.LedgerWalletAdapter(),
    },
    [DefaultWalletType.Solong]: {
        name: "Solong",
        url: "https://solongwallet.com/",
        icon: "https://raw.githubusercontent.com/solana-labs/oyster/main/assets/wallets/solong.png",
        makeAdapter: () => new adapters_1.SolanaWalletAdapter(new wallet_adapter_solong_1.SolongWalletAdapter()),
        isInstalled: () => window.solong !== undefined,
    },
    [DefaultWalletType.Glow]: {
        name: "Glow",
        url: "https://www.glow.app",
        icon: icons_1.GLOW,
        makeAdapter: () => new adapters_1.SolanaWalletAdapter(new wallet_adapter_glow_1.GlowWalletAdapter()),
        isInstalled: () => Boolean(window.glow),
        isMobile: true,
    },
    [DefaultWalletType.Phantom]: {
        name: "Phantom",
        url: "https://www.phantom.app",
        icon: icons_1.PHANTOM,
        makeAdapter: () => new adapters_1.SolanaWalletAdapter(new wallet_adapter_phantom_1.PhantomWalletAdapter()),
        isInstalled: () => { var _a; return ((_a = window.solana) === null || _a === void 0 ? void 0 : _a.isPhantom) === true; },
        isMobile: true,
    },
    [DefaultWalletType.MathWallet]: {
        name: "MathWallet",
        url: "https://www.mathwallet.org",
        icon: icons_1.MATHWALLET,
        makeAdapter: () => new adapters_1.SolanaWalletAdapter(new wallet_adapter_mathwallet_1.MathWalletAdapter()),
        isInstalled: () => { var _a; return ((_a = window.solana) === null || _a === void 0 ? void 0 : _a.isMathWallet) === true; },
        isMobile: true,
    },
    [DefaultWalletType.Coin98]: {
        name: "Coin98",
        url: "https://wallet.coin98.com/",
        icon: icons_1.COIN98,
        makeAdapter: () => new adapters_1.SolanaWalletAdapter(new wallet_adapter_coin98_1.Coin98WalletAdapter()),
        isInstalled: () => window.coin98 !== undefined,
        isMobile: true,
    },
    [DefaultWalletType.Clover]: {
        name: "Clover",
        url: "https://clover.finance",
        icon: icons_1.CLOVER,
        makeAdapter: () => new adapters_1.SolanaWalletAdapter(new wallet_adapter_clover_1.CloverWalletAdapter()),
        isInstalled: () => window.clover_solana !== undefined,
    },
    [DefaultWalletType.SecretKey]: {
        name: "Secret Key",
        url: "https://solana.com/",
        icon: icons_1.FILE,
        makeAdapter: () => new secret_key_1.SecretKeyAdapter(),
    },
    [DefaultWalletType.Solflare]: {
        name: "Solflare (Web)",
        url: "https://solflare.com/provider",
        icon: icons_1.SOLFLARE,
        makeAdapter: (provider, network) => new adapters_1.SolanaWalletAdapter(new solflare_1.SolflareAdapter({
            provider,
            network: network,
        })),
    },
    [DefaultWalletType.SolflareExtension]: {
        name: "Solflare (Extension)",
        url: "https://solflare.com/",
        icon: icons_1.SOLFLARE,
        makeAdapter: () => new adapters_1.SolanaWalletAdapter(new wallet_adapter_solflare_1.SolflareWalletAdapter()),
        isInstalled: () => { var _a; return ((_a = window.solflare) === null || _a === void 0 ? void 0 : _a.isSolflare) === true; },
        isMobile: true,
    },
    [DefaultWalletType.Slope]: {
        name: "Slope",
        url: "https://www.slope.finance/",
        icon: icons_1.SLOPE,
        makeAdapter: () => new adapters_1.SolanaWalletAdapter(new wallet_adapter_slope_1.SlopeWalletAdapter()),
        isInstalled: () => window.Slope !== undefined,
        isMobile: true,
    },
    [DefaultWalletType.Huobi]: {
        name: "HuobiWallet",
        url: "https://www.huobiwallet.io",
        icon: icons_1.HUOBI,
        makeAdapter: () => new adapters_1.SolanaWalletAdapter(new wallet_adapter_huobi_1.HuobiWalletAdapter()),
        isInstalled: () => { var _a; return ((_a = window.huobiWallet) === null || _a === void 0 ? void 0 : _a.isHuobiWallet) === true; },
        isMobile: true,
    },
    [DefaultWalletType.ReadOnly]: {
        name: "Debug",
        url: "https://github.com/saber-hq/saber-common",
        icon: icons_1.MAGNIFYING_GLASS,
        makeAdapter: () => new readonly_1.ReadonlyAdapter(),
        isInstalled: () => !!process.env.LOCAL_PUBKEY || !!process.env.REACT_APP_LOCAL_PUBKEY,
    },
};
