// @cardinal/creator-standard/dist/cjs/generated/accounts/MintManager.js
"use strict";
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.mintManagerBeet = exports.MintManager = void 0;
const tslib_1 = require("tslib");
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
const beet = tslib_1.__importStar(require("@metaplex-foundation/beet"));
const beetSolana = tslib_1.__importStar(require("@metaplex-foundation/beet-solana"));
/**
 * Holds the data for the {@link MintManager} Account and provides de/serialization
 * functionality for that data
 *
 * @category Accounts
 * @category generated
 */
class MintManager {
    constructor(accountType /* size: 8 */, version, mint, authority, ruleset, inUseBy) {
        this.accountType = accountType;
        this.version = version;
        this.mint = mint;
        this.authority = authority;
        this.ruleset = ruleset;
        this.inUseBy = inUseBy;
    }
    /**
     * Creates a {@link MintManager} instance from the provided args.
     */
    static fromArgs(args) {
        return new MintManager(args.accountType, args.version, args.mint, args.authority, args.ruleset, args.inUseBy);
    }
    /**
     * Deserializes the {@link MintManager} from the data of the provided {@link web3.AccountInfo}.
     * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
     */
    static fromAccountInfo(accountInfo, offset = 0) {
        return MintManager.deserialize(accountInfo.data, offset);
    }
    /**
     * Retrieves the account info from the provided address and deserializes
     * the {@link MintManager} from its data.
     *
     * @throws Error if no account info is found at the address or if deserialization fails
     */
    static async fromAccountAddress(connection, address) {
        const accountInfo = await connection.getAccountInfo(address);
        if (accountInfo == null) {
            throw new Error(`Unable to find MintManager account at ${address}`);
        }
        return MintManager.fromAccountInfo(accountInfo, 0)[0];
    }
    /**
     * Provides a {@link web3.Connection.getProgramAccounts} config builder,
     * to fetch accounts matching filters that can be specified via that builder.
     *
     * @param programId - the program that owns the accounts we are filtering
     */
    static gpaBuilder(programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
        return beetSolana.GpaBuilder.fromStruct(programId, exports.mintManagerBeet);
    }
    /**
     * Deserializes the {@link MintManager} from the provided data Buffer.
     * @returns a tuple of the account data and the offset up to which the buffer was read to obtain it.
     */
    static deserialize(buf, offset = 0) {
        return exports.mintManagerBeet.deserialize(buf, offset);
    }
    /**
     * Serializes the {@link MintManager} into a Buffer.
     * @returns a tuple of the created Buffer and the offset up to which the buffer was written to store it.
     */
    serialize() {
        return exports.mintManagerBeet.serialize(this);
    }
    /**
     * Returns the byteSize of a {@link Buffer} holding the serialized data of
     * {@link MintManager} for the provided args.
     *
     * @param args need to be provided since the byte size for this account
     * depends on them
     */
    static byteSize(args) {
        const instance = MintManager.fromArgs(args);
        return exports.mintManagerBeet.toFixedFromValue(instance).byteSize;
    }
    /**
     * Fetches the minimum balance needed to exempt an account holding
     * {@link MintManager} data from rent
     *
     * @param args need to be provided since the byte size for this account
     * depends on them
     * @param connection used to retrieve the rent exemption information
     */
    static async getMinimumBalanceForRentExemption(args, connection, commitment) {
        return connection.getMinimumBalanceForRentExemption(MintManager.byteSize(args), commitment);
    }
    /**
     * Returns a readable version of {@link MintManager} properties
     * and can be used to convert to JSON and/or logging
     */
    pretty() {
        return {
            accountType: this.accountType,
            version: this.version,
            mint: this.mint.toBase58(),
            authority: this.authority.toBase58(),
            ruleset: this.ruleset.toBase58(),
            inUseBy: this.inUseBy,
        };
    }
}
exports.MintManager = MintManager;
/**
 * @category Accounts
 * @category generated
 */
exports.mintManagerBeet = new beet.FixableBeetStruct([
    ['accountType', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['version', beet.u8],
    ['mint', beetSolana.publicKey],
    ['authority', beetSolana.publicKey],
    ['ruleset', beetSolana.publicKey],
    ['inUseBy', beet.coption(beetSolana.publicKey)],
], MintManager.fromArgs, 'MintManager');
