// @cardinal/namespaces-components/dist/cjs/components/ButtonWithFooter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BigButton = exports.FooterText = exports.BottomArea = exports.ButtonWithFooter = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("@emotion/react/jsx-runtime");
const styled_1 = tslib_1.__importDefault(require("@emotion/styled"));
const polished_1 = require("polished");
const LoadingSpinner_1 = require("../common/LoadingSpinner");
const ButtonWithFooter = (_a) => {
    var { footer, children, disabled, loading, complete } = _a, props = tslib_1.__rest(_a, ["footer", "children", "disabled", "loading", "complete"]);
    return ((0, jsx_runtime_1.jsxs)(exports.BottomArea, { children: [(0, jsx_runtime_1.jsx)(exports.BigButton, Object.assign({ disabled: disabled }, props, { children: loading ? ((0, jsx_runtime_1.jsx)(LoadingSpinner_1.LoadingSpinner, {})) : complete ? ((0, jsx_runtime_1.jsx)("i", { className: "fas fa-check-circle" })) : (children) })), (0, jsx_runtime_1.jsx)(exports.FooterText, { children: footer })] }));
};
exports.ButtonWithFooter = ButtonWithFooter;
exports.BottomArea = styled_1.default.div `
  margin-top: 40px;
  left: 28px;
  right: 28px;
  bottom: 28px;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 18px;
`;
exports.FooterText = styled_1.default.div `
  font-size: 12px;
  line-height: 15px;
  letter-spacing: -0.02em;
  color: #696969;
  & > a {
    color: #696969;
    font-weight: bold;
  }
`;
exports.BigButton = styled_1.default.button `
  border: none;
  outline: none;
  border-radius: 16px;
  height: 55px;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  font-weight: bold;
  font-size: 16px;
  line-height: 20px;
  text-align: center;
  background: #000000;
  color: #fff;
  cursor: ${({ disabled }) => (disabled ? 'default' : 'pointer')};

  opacity: ${({ disabled }) => (disabled ? 0.3 : 1)};
  &:hover {
    background: ${({ disabled }) => (disabled ? '' : (0, polished_1.lighten)(0.133, '#000'))};
  }
  &:active {
    background: ${({ disabled }) => (disabled ? '' : (0, polished_1.lighten)(0.212, '#000'))};
  }
`;
