// @cardinal/namespaces-components/dist/cjs/components/ClaimCard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClaimCardOuter = exports.ClaimCard = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("@emotion/react/jsx-runtime");
const styled_1 = tslib_1.__importDefault(require("@emotion/styled"));
const NameEntryClaim_1 = require("./NameEntryClaim");
const ClaimCard = ({ appName, appTwitter, dev, cluster, connection, secondaryConnection, wallet, notify, onComplete, }) => {
    return ((0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: (0, jsx_runtime_1.jsx)(exports.ClaimCardOuter, { children: (0, jsx_runtime_1.jsx)(NameEntryClaim_1.NameEntryClaim, { dev: dev, cluster: cluster, connection: connection !== null && connection !== void 0 ? connection : null, secondaryConnection: secondaryConnection, wallet: wallet !== null && wallet !== void 0 ? wallet : null, appName: appName, appTwitter: appTwitter, notify: notify, onComplete: onComplete }) }) }));
};
exports.ClaimCard = ClaimCard;
exports.ClaimCardOuter = styled_1.default.div `
  width 100%;
  height: 100%;
  position: relative;
  margin: 0px auto;
  min-height: 200px;
  padding: 0px 20px;
`;
