// @cardinal/namespaces-components/dist/cjs/components/Message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("@emotion/react/jsx-runtime");
const styled_1 = tslib_1.__importDefault(require("@emotion/styled"));
const Message = ({ title, description }) => {
    return ((0, jsx_runtime_1.jsx)(Wrapper, { children: (0, jsx_runtime_1.jsxs)(Info, { children: [title && (0, jsx_runtime_1.jsx)(Title, { children: title }), (0, jsx_runtime_1.jsx)(Description, { children: description })] }) }));
};
exports.Message = Message;
const Wrapper = styled_1.default.div `
  background: #e6f7ff;
  border: 1px solid #91d5ff;
  width: 100%;
`;
const Info = styled_1.default.div `
  display: flex;
  flex-direction: column;
  gap: 4px;
`;
const Title = styled_1.default.span `
  font-weight: bold;
  font-size: 14px;
  line-height: 18px;
  letter-spacing: -0.02em;
  color: #000000;
`;
const Description = styled_1.default.p `
  margin: 0;
  font-size: 12px;
  line-height: 15px;
  letter-spacing: -0.02em;
  color: #696969;
`;
