// @cardinal/namespaces-components/dist/cjs/components/PoweredByFooter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PoweredByFooter = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("@emotion/react/jsx-runtime");
const styled_1 = tslib_1.__importDefault(require("@emotion/styled"));
const LogoTitled_1 = require("../common/LogoTitled");
const PoweredByFooter = () => {
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({ style: {
            margin: "0px auto",
            width: "100%",
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
            gap: "6px",
        } }, { children: (0, jsx_runtime_1.jsxs)(StyledLogo, { children: [(0, jsx_runtime_1.jsx)("div", { children: "POWERED BY" }), (0, jsx_runtime_1.jsx)("div", { children: (0, jsx_runtime_1.jsx)(LogoTitled_1.LogoTitled, {}) })] }) })));
};
exports.PoweredByFooter = PoweredByFooter;
const StyledLogo = styled_1.default.div `
  display: flex;
  gap: 5px;
  font-size: 15px;
  svg {
    height: 14px;
  }
`;
