// @cardinal/namespaces-components/dist/cjs/components/ProfileSmall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProfileSmall = void 0;
const jsx_runtime_1 = require("@emotion/react/jsx-runtime");
const hi_1 = require("react-icons/hi");
const format_1 = require("../utils/format");
const AddressImage_1 = require("./AddressImage");
const DisplayAddress_1 = require("./DisplayAddress");
const ProfileSmall = ({ connection, address, dark, onClick, placeholder, className, style, }) => {
    return ((0, jsx_runtime_1.jsxs)("div", Object.assign({ className: `${className} flex cursor-pointer gap-2 text-sm`, style: style, onClick: onClick }, { children: [(0, jsx_runtime_1.jsx)(AddressImage_1.AddressImage, { connection: connection, address: address || undefined, height: "40px", width: "40px", dark: dark, placeholder: placeholder || ((0, jsx_runtime_1.jsx)("div", Object.assign({ style: {
                        color: 'rgb(209, 213, 219)',
                        cursor: 'pointer',
                        overflow: 'hidden',
                        height: '40px',
                        width: '40px',
                    } }, { children: (0, jsx_runtime_1.jsx)(hi_1.HiUserCircle, { style: {
                            position: 'relative',
                            height: '50px',
                            width: '50px',
                            left: '-5px',
                            top: '-5px',
                        }, className: "relative left-[-5px] top-[-5px] h-[50px] w-[50px]" }) }))) }), (0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)("div", Object.assign({ className: `text-${dark ? 'white' : 'black'}` }, { children: (0, jsx_runtime_1.jsx)(DisplayAddress_1.DisplayAddress, { style: { pointerEvents: 'none' }, connection: connection, address: address || undefined, height: "20px", width: "100px", dark: dark }) })), (0, jsx_runtime_1.jsx)("div", Object.assign({ className: "text-sm text-gray-500" }, { children: (0, format_1.shortPubKey)(address) }))] })] })));
};
exports.ProfileSmall = ProfileSmall;
