// @cardinal/namespaces-components/dist/cjs/components/StepDetail.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepDetail = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("@emotion/react/jsx-runtime");
const styled_1 = tslib_1.__importDefault(require("@emotion/styled"));
const StepDetail = ({ icon, title, description, disabled, }) => {
    return ((0, jsx_runtime_1.jsxs)(Wrapper, Object.assign({ disabled: disabled }, { children: [icon, (0, jsx_runtime_1.jsxs)(Info, { children: [(0, jsx_runtime_1.jsx)(Title, { children: title }), (0, jsx_runtime_1.jsx)(Description, { children: description })] })] })));
};
exports.StepDetail = StepDetail;
const Wrapper = styled_1.default.div `
  display: grid;
  grid-template-columns: 18px 1fr;
  grid-column-gap: 9px;
  width: 100%;
  opacity: ${({ disabled }) => (disabled ? 0.3 : 1)};
`;
const Info = styled_1.default.div `
  display: flex;
  flex-direction: column;
  gap: 4px;
`;
const Title = styled_1.default.span `
  font-weight: bold;
  font-size: 14px;
  line-height: 18px;
  letter-spacing: -0.02em;
  color: #000000;
`;
const Description = styled_1.default.div `
  margin: 0;
  font-size: 12px;
  line-height: 15px;
  letter-spacing: -0.02em;
  color: #696969;
`;
