// @cardinal/creator-standard/dist/esm/generated/instructions/Approve.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as splToken from '@solana/spl-token';
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
import { approveIxBeet } from '../types/ApproveIx';
/**
 * @category Instructions
 * @category Approve
 * @category generated
 */
export const ApproveStruct = new beet.BeetArgsStruct([
    ['instructionDiscriminator', beet.u8],
    ['approveIx', approveIxBeet],
], 'ApproveInstructionArgs');
export const approveInstructionDiscriminator = 7;
/**
 * Creates a _Approve_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category Approve
 * @category generated
 */
export function createApproveInstruction(accounts, args, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    var _a;
    const [data] = ApproveStruct.serialize({
        instructionDiscriminator: approveInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.mintManager,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.ruleset,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.mint,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.holderTokenAccount,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.holder,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.delegate,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: (_a = accounts.tokenProgram) !== null && _a !== void 0 ? _a : splToken.TOKEN_PROGRAM_ID,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
