// @cardinal/creator-standard/dist/esm/generated/instructions/CloseRuleset.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
/**
 * @category Instructions
 * @category CloseRuleset
 * @category generated
 */
export const CloseRulesetStruct = new beet.BeetArgsStruct([['instructionDiscriminator', beet.u8]], 'CloseRulesetInstructionArgs');
export const closeRulesetInstructionDiscriminator = 2;
/**
 * Creates a _CloseRuleset_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category CloseRuleset
 * @category generated
 */
export function createCloseRulesetInstruction(accounts, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    const [data] = CloseRulesetStruct.serialize({
        instructionDiscriminator: closeRulesetInstructionDiscriminator,
    });
    const keys = [
        {
            pubkey: accounts.ruleset,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.authority,
            isWritable: true,
            isSigner: true,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
