// @cardinal/creator-standard/dist/esm/generated/instructions/SetInUseBy.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
/**
 * @category Instructions
 * @category SetInUseBy
 * @category generated
 */
export const SetInUseByStruct = new beet.BeetArgsStruct([['instructionDiscriminator', beet.u8]], 'SetInUseByInstructionArgs');
export const setInUseByInstructionDiscriminator = 5;
/**
 * Creates a _SetInUseBy_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @category Instructions
 * @category SetInUseBy
 * @category generated
 */
export function createSetInUseByInstruction(accounts, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    const [data] = SetInUseByStruct.serialize({
        instructionDiscriminator: setInUseByInstructionDiscriminator,
    });
    const keys = [
        {
            pubkey: accounts.mintManager,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.ruleset,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.inUseByAddress,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.holder,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.holderTokenAccount,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
