// @cardinal/creator-standard/dist/esm/generated/instructions/UpdateMintManager.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
import { updateMintManagerIxBeet, } from '../types/UpdateMintManagerIx';
/**
 * @category Instructions
 * @category UpdateMintManager
 * @category generated
 */
export const UpdateMintManagerStruct = new beet.BeetArgsStruct([
    ['instructionDiscriminator', beet.u8],
    ['updateMintManagerIx', updateMintManagerIxBeet],
], 'UpdateMintManagerInstructionArgs');
export const updateMintManagerInstructionDiscriminator = 4;
/**
 * Creates a _UpdateMintManager_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category UpdateMintManager
 * @category generated
 */
export function createUpdateMintManagerInstruction(accounts, args, programId = new web3.PublicKey('creatS3mfzrTGjwuLD1Pa2HXJ1gmq6WXb4ssnwUbJez')) {
    var _a;
    const [data] = UpdateMintManagerStruct.serialize({
        instructionDiscriminator: updateMintManagerInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.mintManager,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.ruleset,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.authority,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: accounts.payer,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: (_a = accounts.systemProgram) !== null && _a !== void 0 ? _a : web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
